File: client/src/main/scala/stasis/client/tracking/state/serdes/RecoveryStateSerdes.scala
package stasis.client.tracking.state.serdes

import stasis.client.model.proto
import stasis.client.tracking.state.RecoveryState
import stasis.core.persistence.backends.file.state.StateStore
import stasis.shared.ops.Operation

import java.util.UUID
import scala.util.Try

object RecoveryStateSerdes {
  implicit val recoverySerdes: StateStore.Serdes[Map[Operation.Id, RecoveryState]] =
    new StateStore.Serdes[Map[Operation.Id, RecoveryState]] {
      override def serialize(state: Map[Operation.Id, RecoveryState]): Array[Byte] =
        proto.state
          .RecoveryStateCollection(
            collection = state.map { case (k, v) => k.toString -> RecoveryState.toProto(state = v) }
          )
          .toByteArray

      override def deserialize(bytes: Array[Byte]): Try[Map[Operation.Id, RecoveryState]] =
        proto.state.RecoveryStateCollection
          .parseFrom(bytes)
          .collection
          .foldLeft(Try(Seq.empty[RecoveryState])) { case (collected, (k, v)) =>
            for {
              collected <- collected
              state <- RecoveryState.fromProto(operation = UUID.fromString(k), state = v)
            } yield {
              collected :+ state
            }
          }
          .map(_.map(state => state.operation -> state).toMap)
    }
}

File: client-android/lib/src/main/kotlin/stasis/client_android/lib/tracking/state/serdes/RecoveryStateSerdes.kt
package stasis.client_android.lib.tracking.state.serdes

import stasis.client_android.lib.ops.OperationId
import stasis.client_android.lib.persistence.state.StateStore
import stasis.client_android.lib.tracking.state.RecoveryState
import stasis.client_android.lib.utils.Try
import java.util.UUID

object RecoveryStateSerdes : StateStore.Serdes<Map<OperationId, RecoveryState>> {
    override fun serialize(state: Map<OperationId, RecoveryState>): ByteArray =
        stasis.client_android.lib.model.proto.RecoveryStateCollection(
            collection = state.map { (k, v) -> k.toString() to RecoveryState.toProto(state = v) }.toMap()
        ).encode()

    override fun deserialize(bytes: ByteArray): Try<Map<OperationId, RecoveryState>> = Try {
        stasis.client_android.lib.model.proto.RecoveryStateCollection.ADAPTER
            .decode(bytes)
            .collection
            .map { (k, v) ->
                val operation = UUID.fromString(k)
                operation to RecoveryState.fromProto(operation = operation, state = v).get()
            }.toMap()
    }
}
